src/nanoemoji/nanoemoji.py

```diff
@@ -270,16 +272,30 @@ def _draw_glyph_extents(ufo: ufoLib2.Font, glyph: Glyph):
 
 
 def _glyf_ufo(ufo, color_glyphs):
+    # glyphs by reuse_key
+    glyphs = {}
+    swaps = []
     for color_glyph in color_glyphs:
-        svg_units_to_font_units = color_glyph.transform_for_font_space()
         logging.debug(
             "%s %s %s",
             ufo.info.familyName,
             color_glyph.glyph_name,
-            svg_units_to_font_units,
+            color_glyph.transform_for_font_space(),
         )
+        parent_glyph = ufo.get(color_glyph.glyph_name)
         for painted_layer in color_glyph.as_painted_layers():
-            _create_glyph(color_glyph, painted_layer)
+            # if we've seen this shape before reuse it
+            reuse_key = _inter_glyph_reuse_key(painted_layer)
+            if reuse_key not in glyphs:
+                glyph = _create_glyph(color_glyph, painted_layer)
+                glyphs[reuse_key] = glyph
+            else:
+                glyph = glyphs[reuse_key]
+            parent_glyph.components.append(Component(baseGlyph=glyph.name))
+
+        # No great reason to keep single-component glyphs around
```

Comment:
it's that single-component glyphs are wasteful, but rather the fact that a given glyph is only ever referenced as a component in only one other composite glyph (and is not itself a primary glyph with a direct cmap or GSUB mapping, but it's only reason of existence is that it is used as component).
E.g. if glyph C is referenced by both A and B, even if the latter two only have one component each, you're still reusing C in both, thus saving a bit of space. But if glyph C (and D, E, F, etc.) is (are) only used as component(s) in _one_ other glyph, say, A, then you are wasting space to keep them as individual glyphs and A may better be a simple, decomposed glyph.